Adds the Intento!

Fikou opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
Adds the Intento toy! It gives you intents to click on which you follow! It has database integration with high-scores.
Video here: https://streamable.com/tr2e17

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
A fun toy for the whole family!

## Changelog
:cl: Fikou, Floyd, coiax
add: Adds the Intento! Buy it in your local library vendor!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Fikou committed:
Adds back intents

Fikou committed:
achievement

Fikou committed:
adds to vendor

Fikou committed:
my stomach's riddled with holes

Bluedazzled commented:
PR: [tgstation/tgstation] Pull Request opened by Fikou: #56770 Fikou - Adds Back Intents
you filthy LIAR

Paxilmaniac commented:
![image](https://user-images.githubusercontent.com/57122098/107300795-312aa500-6a48-11eb-9355-04c17111d695.png)
One bait and switch please, thank you.

EOBGames commented:
High quality bait, and a high quality toy, too

JohnFulpWillard commented:
we dont have permission to straight up use the word 'Nintendo' for a gaming console, do we?

Jared-Fogle commented:
"nintento" is not the fair use parody you think it is

EOBGames commented:
> we dont have permission to straight up use the word 'Nintendo' for a gaming console, do we?

I think you'll find it's clearly called a "Nintento", totally original, not related to any existing video game company.

Jared-Fogle commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,98 @@
+	var/usable = FALSE
+	/// Associated list of intents to their sounds
+	var/static/list/sound_by_intent = list(HELP = 'sound/items/intents/Help.ogg',
+										DISARM = 'sound/items/intents/Disarm.ogg',
+										GRAB = 'sound/items/intents/Grab.ogg',
+										HARM = 'sound/items/intents/Harm.ogg',
+										)
```
use normal indentation you maniac

Jared-Fogle commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,98 @@
+	var/input
+	var/icon_x = text2num(modifiers["icon-x"])
+	var/icon_y = text2num(modifiers["icon-y"])
+	if(icon_x > 16 && icon_y > 16)
+		input = DISARM
+	if(icon_x < 16 && icon_y > 16)
+		input = HELP
+	if(icon_x > 16 && icon_y < 16)
+		input = GRAB
+	if(icon_x < 16 && icon_y < 16)
+		input = HARM
```
make the 16 a define

Qustinnus replied:
or just do world.icon_size/2

Fikou commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,98 @@
+	player_sequence.Cut()
+	current_sequence += pick(list(HELP, DISARM, GRAB, HARM))
+	for(var/input in current_sequence)
+		sleep(0.6 SECONDS)
```
some better way to do this please

Jared-Fogle replied:
process

MrMelbert commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,98 @@
+
+/obj/item/toy/nintento
+	name = "Nintento"
+	desc = "Fundamentally useless for all intensive purposes."
```
```suggestion
	desc = "Fundamentally useless for all intent-sive purposes."
```
![image](https://user-images.githubusercontent.com/51863163/107301522-405a2480-6a41-11eb-85e0-d5ec6fc3ac70.png)

MrDoomBringer replied:
woosh amirite

MrDoomBringer commented:
![image](https://user-images.githubusercontent.com/29008542/107302042-abf4bf80-6a4a-11eb-8b43-76d131bd794c.png)
ugh im terrible

Jared-Fogle commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,98 @@
+/obj/item/toy/nintento/proc/render(input)
+	icon_state = input
+	playsound(src, sound_by_intent[input], 50, FALSE)
+	addtimer(VARSET_CALLBACK(src, icon_state, initial(icon_state)), 0.5 SECONDS)
```
no, use processing for this. short timers that repeat kill the timer ss

Mothblocks reviewed (changes requested):
be at least a little creative with the name

i get the pun but it is weak

Fikou committed:
no fun allowed

Jared-Fogle commented:
good enough

cacogen commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,100 @@
+	usable = TRUE
+
+/obj/item/toy/intento/proc/start()
+	say("Game Starting!")
```
```suggestion
	say("Game starting!")
```

Fikou committed:
sure

itseasytosee commented:
Emag effect?

Jared-Fogle commented:
it should make it so whatever it does it does on you so if its harm the toy punches you

cacogen commented:
yes, punching should be an atom proc

Fikou committed:
emag

Jared-Fogle commented on code/game/objects/items/toys.dm:
```diff
@@ -1630,6 +1639,27 @@
+	ADD_TRAIT(src, TRAIT_NODROP, "intento")
+	to_chat(victim, "<span class='userdanger'>Bad mistake.</span>")
+	for(var/input in combo_inputs)
+		sleep(0.6 SECONDS)
```
🥴

Fikou replied:
booba

coiax reviewed (commented):
Also, shouldn't you fail a round if you don't take an action after a certain amount of time?

ATH1909 commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,130 @@
+
+/obj/item/toy/intento/proc/retaliate(list/combo_inputs, mob/living/victim)
+	ADD_TRAIT(src, TRAIT_NODROP, "intento")
+	to_chat(victim, "<span class='userdanger'>Bad mistake.</span>")
```
make the toy actually say this out loud so that it shows up in runechat

Fikou replied:
no

ATH1909 commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,130 @@
+		switch(input)
+			if(HELP)
+				to_chat(victim, "<span class='danger'>[src] hugs you to make you feel better!</span>")
+				victim.Dizzy(20)
```
I kind of like the idea of the help parts of the combo just giving you a positive hug moodlet and not actually debuffing you; would you mind replacing the Dizzy() part of the combo here with that?

also, IIRC, Dizzy uses seconds, not deciseconds, so this lasts 10x as long as most of these other effects

Fikou replied:
sure

ATH1909 commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,130 @@
+				victim.Dizzy(20)
+			if(DISARM)
+				to_chat(victim, "<span class='danger'>You're knocked down from a shove by [src]!</span>")
+				victim.Knockdown(20)
```
make this last the actual shove knockdown duration if you're not paralyzed or knocked down, paralyze you for the actual shovestun duration if you're already knocked down but not yet paralyzed, or do nothing (or maybe a throw or something?) if you're already paralyzed.

also maybe consider making it apply the shove debuff (and gun disarming if you REALLY want to go all into mimicking a real disarm), but that's just if you want to go the extra mile

Fikou replied:
no

ATH1909 commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,130 @@
+				victim.Knockdown(20)
+			if(GRAB)
+				to_chat(victim, "<span class='danger'>[src] grabs you aggressively!</span>")
+				victim.Stun(20)
```
```suggestion
				victim.Immobilize(20)
```

ATH1909 commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,130 @@
+				victim.Knockdown(20)
+			if(GRAB)
+				to_chat(victim, "<span class='danger'>[src] grabs you aggressively!</span>")
+				victim.Stun(20)
```
on the one hand, I'd prefer to see an Immobilize() used here, but on the other, Stun is more accurate to an aggressive grab

if you wanted to go super in-depth, you could make this apply some losebreath if you're already stunned (to simulate the toy escalating all the way up to a kill grab (skipping a neck grab, I guess)).

if you wanted to go ultra god-tier PR with this, you could make this slam you into an adjacent table if there's one near you and you're already stunned.

Fikou replied:
a

ATH1909 commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,130 @@
+				victim.Stun(20)
+			if(HARM)
+				to_chat(victim, "<span class='danger'>You're punched by [src]!</span>")
+				victim.apply_damage(rand(20, 30), BRUTE)
```
add a 10% chance for a knockdown here, like what the punches of most roundstart species get

Fikou replied:
me

cacogen commented:
It needs sound effects for the attacks it does

MrDoomBringer commented:
@cacogen it already does you goober

it's handled in the `render(intent)` proc

Fikou commented on code/game/objects/items/toys.dm:
```diff
@@ -1536,3 +1538,130 @@
+		render(input)
+		switch(input)
+			if(HELP)
+				to_chat(victim, "<span class='danger'>[src] hugs you to make you feel better!</span>")
+				victim.Dizzy(20)
+			if(DISARM)
+				to_chat(victim, "<span class='danger'>You're knocked down from a shove by [src]!</span>")
+				victim.Knockdown(20)
+			if(GRAB)
+				to_chat(victim, "<span class='danger'>[src] grabs you aggressively!</span>")
+				victim.Stun(20)
+			if(HARM)
+				to_chat(victim, "<span class='danger'>You're punched by [src]!</span>")
```
good idea

Mothblocks reviewed (changes requested):
Use process, not sleeps.

coiax committed:
Modify Intento to use fast processing

Instead of using the timer subsystem and sleeps, the Intento now uses
the fast processing system.

The Intento is in one of a number of different states, some of which
will have accompanying processing with SSfastprocess. In addition, there
is a separate internal timer for resetting the icon state after it
flashes an icon.

- The various constant times, such as flashing between icons in the
  demonstration state, or waiting for it to start up, are now defines.

Fikou committed:
Merge pull request #4 from coiax/process-for-intento

Modify Intento to use fast processing

Fikou committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into i-feel-a-bit-queasy

Mothblocks commented on code/game/objects/items/toys.dm:
```diff
@@ -1593,75 +1613,166 @@
+
+	state = STATE_STARTING
+	next_process = world.time + TIME_TO_BEGIN
+	START_PROCESSING(SSfastprocess, src)
```
Make sure you STOP_PROCESSING at some point too, like in Destroy.

Mothblocks commented on code/game/objects/items/toys.dm:
```diff
@@ -1593,75 +1613,166 @@
-		addtimer(CALLBACK(src, .proc/start_round, 1 SECONDS))
+
+		state = STATE_STARTING
+		next_process = world.time + TIME_TO_BEGIN
```
Use the cooldown defines.

Mothblocks replied:
Or use `next_process = TIME_TO_BEGIN` then `next_process -= delta_time`, and check if `next_process <= 0`.

Mothblocks commented on code/game/objects/items/toys.dm:
```diff
@@ -1593,75 +1613,166 @@
+			process_demo()
+
+		if(STATE_END_OF_GAME)
+			var/mob/living/holder = null
+			if(isliving(loc))
+				holder = loc
+
+			process_end(holder)
```
```suggestion
			process_end(isliving(loc) ? loc : null)
```

Mothblocks commented on code/game/objects/items/toys.dm:
```diff
@@ -1593,75 +1613,166 @@
+		var/award_status = user.client.get_award_status(/datum/award/score/intento_score)
+		if(award_score - award_status > 0)
+			award_score -= award_status
+		user.client.give_award(/datum/award/score/intento_score, user, award_score, TRUE)
```
```suggestion
		user.client.give_award(/datum/award/score/intento_score, user, award_score)
```

I can't find a TRUE being needed in any definition of give_award. Correct me if I'm wrong.

Fikou committed:
Update code/game/objects/items/toys.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Fikou committed:
Update code/game/objects/items/toys.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

Fikou committed:
yeah

cacogen commented:
iNtento

N like Nanotrasen
Nintendo more clear
iDevice

Fikou committed:
DESTROY

Megamaw commented:
aaah, i'm too late to suggest playing this minigame increase your gaming skill huh

MrDoomBringer commented:
@Megamaw now you can code it and take all the github green squares for yourself!

maxymax13 commented:
emagging it should turn it into the Combato
![combato](https://user-images.githubusercontent.com/38844529/109055935-91634d00-76be-11eb-84f3-93aa360d09b2.png)
